File: mega_apps/l10n_co_pos_receipt_cufe/models/pos_order.py
# -*- coding: utf-8 -*-
# Part of Odoo. See LICENSE file for full copyright and licensing details.

import logging
from odoo import models, fields, api

_logger = logging.getLogger(__name__)


class PosOrder(models.Model):
    _inherit = "pos.order"

    # Campo computado para obtener el CUFE de la factura
    l10n_co_cufe = fields.Char(
        string="CUFE",
        compute="_compute_l10n_co_cufe",
        store=True,
        help="CUFE de la factura electrónica asociada"
    )

    # Campo para determinar si mostrar CUFE en tirilla
    l10n_co_show_cufe_on_receipt = fields.Boolean(
        string="Mostrar CUFE en tirilla",
        compute="_compute_l10n_co_show_cufe_on_receipt"
    )

    @api.depends('account_move', 'account_move.l10n_co_edi_cufe_cude_ref')
    def _compute_l10n_co_cufe(self):
        """Obtiene el CUFE de la factura asociada si existe"""
        for order in self:
            if order.account_move and order.account_move.l10n_co_edi_cufe_cude_ref:
                order.l10n_co_cufe = order.account_move.l10n_co_edi_cufe_cude_ref
            else:
                order.l10n_co_cufe = False

    @api.depends('account_move', 'account_move.l10n_co_dian_state', 'config_id.l10n_co_show_cufe_receipt')
    def _compute_l10n_co_show_cufe_on_receipt(self):
        """Determina si se debe mostrar el CUFE en la tirilla"""
        for order in self:
            order.l10n_co_show_cufe_on_receipt = (
                order.config_id.l10n_co_show_cufe_receipt
                and order.account_move
                and order.account_move.l10n_co_dian_state == 'invoice_accepted'
                and order.account_move.l10n_co_edi_cufe_cude_ref
            )

    def _generate_pos_order_invoice(self):
        """
        Override para enviar automáticamente a DIAN si está configurado
        """
        result = super()._generate_pos_order_invoice()

        # Si está habilitada la facturación automática para Colombia
        if self.config_id.l10n_co_auto_invoice_pos and self.company_id.country_id.code == 'CO':
            if self.account_move and self.account_move.l10n_co_dian_is_enabled:
                try:
                    _logger.info(f"Enviando factura {self.account_move.name} a DIAN automáticamente")
                    self.account_move.l10n_co_dian_action_send_bill_support_document()
                except Exception as e:
                    # Log error pero no bloquear el flujo
                    _logger.warning(f"Error al enviar factura a DIAN para orden {self.name}: {e}")

        return result

File: mega_apps/l10n_co_pos_receipt_cufe/static/src/overrides/models/pos_order.js
import { PosOrder } from "@point_of_sale/app/models/pos_order";
import { patch } from "@web/core/utils/patch";
import { qrCodeSrc } from "@point_of_sale/utils";

patch(PosOrder.prototype, {
    /**
     * Verifica si la orden es de Colombia
     */
    is_colombian_country() {
        return this.company.country_id?.code === "CO";
    },

    /**
     * Extiende export_for_printing para agregar información del CUFE
     */
    export_for_printing(baseUrl, headerData) {
        const result = super.export_for_printing(...arguments);

        // Solo para órdenes colombianas
        if (this.is_colombian_country()) {
            // Agregar nombre de orden (compatibilidad con l10n_co_pos)
            result.l10n_co_dian = this.name;

            // Agregar CUFE si la orden tiene factura asociada
            const hasCufe = this.account_move?.l10n_co_edi_cufe_cude_ref;
            const showCufe = this.config.l10n_co_show_cufe_receipt;
            const isDianAccepted = this.account_move?.l10n_co_dian_state === 'invoice_accepted';

            if (hasCufe && showCufe && isDianAccepted) {
                result.l10n_co_cufe = this.account_move.l10n_co_edi_cufe_cude_ref;
                result.l10n_co_cufe_qr = qrCodeSrc(result.l10n_co_cufe, { size: 180 });
                result.l10n_co_show_cufe = true;
            } else {
                result.l10n_co_cufe = false;
                result.l10n_co_cufe_qr = false;
                result.l10n_co_show_cufe = false;
            }
        }

        return result;
    },

    /**
     * Determina si debe esperar a que se procese la orden
     * (útil si queremos facturar automáticamente)
     */
    wait_for_push_order() {
        let result = super.wait_for_push_order(...arguments);
        // Esperar si está configurado para facturación automática en Colombia
        if (this.config.l10n_co_auto_invoice_pos && this.is_colombian_country()) {
            result = true;
        }
        return result;
    },
});
